source/parser/independent/intermediate-macro-instruction.cpp

```diff
@@ -0,0 +1,148 @@
+/* C++ Assembler Interpreter
+ * Copyright (C) 2016 Chair of Computer Architecture
+ * at Technical University of Munich
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "parser/independent/intermediate-macro-instruction.hpp"
+
+#include "parser/common/compile-error-list.hpp"
+#include "parser/independent/intermediate-instruction.hpp"
+#include "parser/independent/macro-directive-table.hpp"
+#include "parser/independent/macro-directive.hpp"
+#include "parser/independent/memory-allocator.hpp"
+#include "parser/independent/symbol-graph.hpp"
+#include "parser/independent/symbol-replacer.hpp"
+
+
+void IntermediateMacroInstruction::replaceWithMacros(
+    IntermediateOperationVectorIterator begin,
+    IntermediateOperationVectorIterator end,
+    MacroDirectiveTable& macroTable,
+    CompileErrorList& errors) {
+  for (auto i = begin; i != end; ++i) {
+    // Try casting to IntermediateInstruction, skip instruction if it doesn't
+    // work
+    if ((*i)->getType() != IntermediateOperation::Type::INSTRUCTION) continue;
+
+    IntermediateInstruction& inst = static_cast<IntermediateInstruction&>(**i);
```

Comment:
I would normally have said `dynamic_cast`, but since you have the check above it's fine. However, this is a perfect use case for `auto`, since it's clear the result of casting to `IntermediateInstruction&` is `IntermediateInstruction&`